File: emulinker/src/main/java/org/emulinker/net/UDPServer.java
package org.emulinker.net;

import com.codahale.metrics.MetricRegistry;
import com.codahale.metrics.Timer;
import com.google.common.flogger.FluentLogger;
import java.io.IOException;
import java.net.*;
import java.nio.Buffer;
import java.nio.ByteBuffer;
import java.nio.channels.DatagramChannel;
import org.emulinker.util.*;

public abstract class UDPServer implements Executable {
  private static final FluentLogger logger = FluentLogger.forEnclosingClass();
  /*
  	private static int		artificalPacketLossPercentage = 0;
  	private static int		artificalDelay = 0;
  	private static Random	random = new Random();

  	static
  	{
  		try
  		{
  			artificalPacketLossPercentage = Integer.parseInt(System.getProperty("artificalPacketLossPercentage"));
  			artificalDelay = Integer.parseInt(System.getProperty("artificalDelay"));
  		}
  		catch(Exception e) {}

  		if(artificalPacketLossPercentage > 0)
  			logger.atWarning().log("Introducing " + artificalPacketLossPercentage + "% artifical packet loss!");

  		if(artificalDelay > 0)
  			logger.atWarning().log("Introducing " + artificalDelay + "ms artifical delay!");
  	}
  */
  private int bindPort;
  private DatagramChannel channel;
  private boolean isRunning = false;
  private boolean stopFlag = false;

  protected final Timer serverToClientRequests;

  public UDPServer(boolean shutdownOnExit, MetricRegistry metrics) {
    if (shutdownOnExit) Runtime.getRuntime().addShutdownHook(new ShutdownThread());

    this.serverToClientRequests =
        metrics.timer(MetricRegistry.name(UDPServer.class, "clientToServerRequests"));
  }

  public int getBindPort() {
    return bindPort;
  }

  @Override
  public boolean isRunning() {
    return isRunning;
  }

  public synchronized boolean isBound() {
    if (channel == null) return false;
    if (channel.socket() == null) return false;
    return !channel.socket().isClosed();
  }

  public boolean isConnected() {
    return channel.isConnected();
  }

  public synchronized void start() {
    logger.atFine().log(toString() + " received start request!");
    if (isRunning) {
      logger.atFine().log(toString() + " start request ignored: already running!");
      return;
    }

    stopFlag = false;
  }

  protected boolean getStopFlag() {
    return stopFlag;
  }

  @Override
  public synchronized void stop() {
    stopFlag = true;

    if (channel != null) {
      try {
        channel.close();
      } catch (IOException e) {
        logger.atSevere().withCause(e).log("Failed to close DatagramChannel");
      }
    }
  }

  protected synchronized void bind() throws BindException {
    bind(-1);
  }

  protected synchronized void bind(int port) throws BindException {
    try {
      channel = DatagramChannel.open();

      if (port > 0) channel.socket().bind(new InetSocketAddress(port));
      else channel.socket().bind(null);

      bindPort = channel.socket().getLocalPort();

      ByteBuffer tempBuffer = getBuffer();
      int bufferSize = (tempBuffer.capacity() * 2);
      releaseBuffer(tempBuffer);

      channel.socket().setReceiveBufferSize(bufferSize);
      channel.socket().setSendBufferSize(bufferSize);
    } catch (IOException e) {
      throw new BindException("Failed to bind to port " + port, port, e);
    }

    this.start();
  }

  protected abstract ByteBuffer getBuffer();

  protected abstract void releaseBuffer(ByteBuffer buffer);

  protected abstract void handleReceived(ByteBuffer buffer, InetSocketAddress remoteSocketAddress);

  protected void send(ByteBuffer buffer, InetSocketAddress toSocketAddress) {
    if (!isBound()) {
      logger.atWarning().log(
          "Failed to send to "
              + EmuUtil.formatSocketAddress(toSocketAddress)
              + ": UDPServer is not bound!");
      return;
    }
    /*
    if(artificalPacketLossPercentage > 0 && Math.abs(random.nextInt()%100) < artificalPacketLossPercentage)
    {
    	return;
    }
    */
    try {
      //			logger.atFine().log("send("+EmuUtil.dumpBuffer(buffer, false)+")");
      channel.send(buffer, toSocketAddress);
    } catch (Exception e) {
      logger.atSevere().withCause(e).log("Failed to send on port " + getBindPort());
    }
  }

  @Override
  public void run() {
    isRunning = true;
    logger.atFine().log(toString() + ": thread running...");

    try {
      while (!stopFlag) {
        try {
          ByteBuffer buffer = getBuffer();
          InetSocketAddress fromSocketAddress = (InetSocketAddress) channel.receive(buffer);

          if (stopFlag) break;

          if (fromSocketAddress == null)
            throw new IOException(
                "Failed to receive from DatagramChannel: fromSocketAddress == null");
          /*
          if(artificalPacketLossPercentage > 0 && Math.abs(random.nextInt()%100) < artificalPacketLossPercentage)
          {
          	releaseBuffer(buffer);
          	continue;
          }

          if(artificalDelay > 0)
          {
          	try
          	{
          		Thread.sleep(artificalDelay);
          	}
          	catch(Exception e) {}
          }
          */
          // Cast to avoid issue with java version mismatch:
          // https://stackoverflow.com/a/61267496/2875073
          ((Buffer) buffer).flip();
          //					logger.atFine().log("receive("+EmuUtil.dumpBuffer(buffer, false)+")");
          // TODO(nue): time this
          handleReceived(buffer, fromSocketAddress);
          releaseBuffer(buffer);
        } catch (SocketException e) {
          if (stopFlag) break;

          logger.atSevere().withCause(e).log("Failed to receive on port %d", getBindPort());
        } catch (IOException e) {
          if (stopFlag) break;

          logger.atSevere().withCause(e).log("Failed to receive on port %d", getBindPort());
        }
      }
    } catch (Throwable e) {
      logger.atSevere().withCause(e).log(
          "UDPServer on port %d caught unexpected exception!", getBindPort());
      stop();
    } finally {
      isRunning = false;
      logger.atFine().log(toString() + ": thread exiting...");
    }
  }

  private class ShutdownThread extends Thread {
    private ShutdownThread() {}

    @Override
    public void run() {
      UDPServer.this.stop();
    }
  }
}

File: emulinker/src/main/java/org/emulinker/net/UDPServer.kt
package org.emulinker.net

import com.google.common.flogger.FluentLogger
import io.ktor.network.sockets.*
import io.ktor.utils.io.core.*
import java.net.InetSocketAddress
import java.nio.ByteBuffer
import kotlin.Exception
import kotlin.coroutines.CoroutineContext
import kotlinx.coroutines.*
import org.emulinker.kaillera.controller.v086.V086Utils
import org.emulinker.kaillera.controller.v086.V086Utils.toKtorAddress
import org.emulinker.util.EmuUtil.dumpBufferFromBeginning
import org.emulinker.util.EmuUtil.formatSocketAddress
import org.emulinker.util.Executable

private val logger = FluentLogger.forEnclosingClass()

abstract class UDPServer : Executable {
  abstract val bufferSize: Int

  /*
  	private static int		artificalPacketLossPercentage = 0;
  	private static int		artificalDelay = 0;
  	private static Random	random = new Random();

  	static
  	{
  		try
  		{
  			artificalPacketLossPercentage = Integer.parseInt(System.getProperty("artificalPacketLossPercentage"));
  			artificalDelay = Integer.parseInt(System.getProperty("artificalDelay"));
  		}
  		catch(Exception e) {}

  		if(artificalPacketLossPercentage > 0)
  			logger.atWarning().log("Introducing " + artificalPacketLossPercentage + "% artifical packet loss!");

  		if(artificalDelay > 0)
  			logger.atWarning().log("Introducing " + artificalDelay + "ms artifical delay!");
  	}
  */
  // TODO(nue): This is supposed to be the server's port i guess but is always 0..
  val bindPort = 0

  private lateinit var serverSocket: BoundDatagramSocket

  protected lateinit var globalContext: CoroutineContext

  final override var threadIsActive = false
    private set

  protected var stopFlag = false
    private set

  @get:Synchronized
  val isBound: Boolean
    get() {
      return !serverSocket.isClosed
    }

  @Synchronized
  open suspend fun start(udpSocketProvider: UdpSocketProvider, globalContext: CoroutineContext) {
    this.globalContext = globalContext
    logger.atFine().log(toString() + " received start request!")
    if (threadIsActive) {
      logger.atFine().log(toString() + " start request ignored: already running!")
      return
    }
    stopFlag = false
  }

  @Synchronized
  override suspend fun stop() {
    stopFlag = true
    serverSocket.dispose()
  }

  @Synchronized
  protected open fun bind(udpSocketProvider: UdpSocketProvider, port: Int) {
    serverSocket =
        udpSocketProvider.bindSocket(
            io.ktor.network.sockets.InetSocketAddress("0.0.0.0", port), bufferSize)

    logger.atInfo().log("Accepting messages at ${serverSocket.localAddress}")
  }

  protected abstract fun allocateBuffer(): ByteBuffer

  protected abstract suspend fun handleReceived(
      buffer: ByteBuffer, remoteSocketAddress: InetSocketAddress, requestScope: CoroutineScope
  )

  protected suspend fun send(buffer: ByteBuffer, toSocketAddress: InetSocketAddress) {
    if (!isBound) {
      logger
          .atWarning()
          .log(
              "Failed to send to %s: UDPServer is not bound!", formatSocketAddress(toSocketAddress))
      return
    }
    /*
    if(artificalPacketLossPercentage > 0 && Math.abs(random.nextInt()%100) < artificalPacketLossPercentage)
    {
    	return;
    }
    */ try {
      //			logger.atFine().log("send("+EmuUtil.INSTANCE.dumpBuffer(buffer, false)+")");
      //      channel!!.send(buffer, toSocketAddress)
      serverSocket.send(Datagram(ByteReadPacket(buffer), toSocketAddress.toKtorAddress()))
    } catch (e: Exception) {
      logger.atSevere().withCause(e).log("Failed to send on port $bindPort")
    }
  }

  override suspend fun run(globalContext: CoroutineContext) {
    this.globalContext = globalContext
    threadIsActive = true

    while (!stopFlag) {
      supervisorScope {
        val datagram = serverSocket.incoming.receive()

        require(datagram.address is io.ktor.network.sockets.InetSocketAddress) {
          "address was an incompatable type!"
        }

        val buffer = datagram.packet.readByteBuffer()

        val requestContext =
            CoroutineScope(coroutineContext) // TODO(nue): Can we just pass coroutineContext?
        try {
          handleReceived(
              buffer,
              V086Utils.toJavaAddress(
                  datagram.address as io.ktor.network.sockets.InetSocketAddress),
              requestScope = requestContext)
        } catch (e: Exception) {
          if (e is CancellationException) {
            throw e
          }
          logger
              .atSevere()
              .withCause(e)
              .log("Error while handling request: %s", dumpBufferFromBeginning(buffer))
        }
      }
    }

    threadIsActive = false
  }

  // TODO(nue): Investigate this.
  //  private inner class ShutdownThread : Thread() {
  //    override fun run() {
  //      this@UDPServer.stop()
  //    }
  //  }
  //  init {
  //    if (shutdownOnExit) Runtime.getRuntime().addShutdownHook(ShutdownThread())
  //  }
}
